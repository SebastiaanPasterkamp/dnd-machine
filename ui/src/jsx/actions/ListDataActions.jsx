import React from 'react';
import Reflux from 'reflux';
import _ from 'lodash';

import ReportingActions from './ReportingActions.jsx';

function jsonOrBust(response) {
    if (response.ok) {
        return response.json();
    }
    throw response.statusText;
};

var ListDataActions = Reflux.createActions({
    "setState": {},
    "doLogin": {children: ['completed', 'failed']},
    "doLogout": {children: ['completed', 'failed']},
    "fetchItems": {children: ['completed', 'failed']},
});
let throttledGet = {};

ListDataActions.fetchItems.listen((type, category=null) => {
    let path = '/' + _.filter([
        category, type, category ? 'api' : null]).join('/');

    if (!(path in throttledGet)) {
        throttledGet[path] = _.throttle((path, type) => {

            fetch(path, {
                credentials: 'same-origin',
                'headers': {
                    'X-Requested-With': 'XMLHttpRequest'
                }
            })
            .then(jsonOrBust)
            .then((response) => {
                ListDataActions.fetchItems.completed({
                    [type]: response
                });
            })
            .catch((error) => {
                console.log(error);
                ListDataActions.fetchItems.failed(type, error);
            });
        }, 1000, {leading: true, trailing: false});
    }

    throttledGet[path](path, type);
});

ListDataActions.doLogin.listen((credentials, success, failure) => {
    fetch('/login', {
        credentials: 'same-origin',
        method: 'POST',
        'headers': {
            'X-Requested-With': 'XMLHttpRequest',
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(credentials)
    })
    .then(jsonOrBust)
    .then((response) => {
        ListDataActions.doLogin.completed(response);
        ListDataActions.fetchItems.completed({
            current_user: response
        });

        ReportingActions.showMessage(
            'good',
            'Login successful',
            'Login',
            5
        );

        if (success) {
            success(content);
        }
    })
    .catch((error) => {
        console.log(error);

        ReportingActions.showMessage(
            'bad',
            'Login failed',
            'Login',
            10
        );

        ListDataActions.doLogin.failed(error);

        if (failure) {
            failure();
        }
    });
});

ListDataActions.doLogout.listen((callback) => {
    fetch('/logout', {
        credentials: 'same-origin',
        'headers': {
            'X-Requested-With': 'XMLHttpRequest'
        }
    })
    .then(jsonOrBust)
    .then((response) => {
        ListDataActions.doLogout.completed(response);
        ListDataActions.fetchItems.completed({
            current_user: null
        });

        ReportingActions.showMessage(
            'good',
            'Logout successful',
            'Logout',
            5
        );

        if (callback) {
            callback();
        }
    })
    .catch((error) => {
        console.log(error);
        ListDataActions.doLogout.failed(error);
        ListDataActions.fetchItems.completed({
            current_user: null
        });

        ReportingActions.showMessage(
            'bad',
            'Logout failed',
            'Logout',
            10
        );

        if (callback) {
            callback();
        }
    });
});

export default ListDataActions;
