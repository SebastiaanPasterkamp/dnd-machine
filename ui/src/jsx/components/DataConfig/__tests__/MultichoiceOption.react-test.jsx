import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import ListDataActions from 'actions/ListDataActions';
import {
    mockedApi,
} from '../../../../../tests/__mocks__';

import { MultichoiceOption } from '..';
import uuidv4 from '../utils/uuidv4';
jest.mock('../utils/uuidv4');

const mockedUUID1 = 'mocked-uuid-1';
const mockedUUID2 = 'mocked-uuid-2';

describe('Component: MultichoiceOption', () => {
    const fullProps = {
        type: 'multichoice',
        uuid: mockedUUID1,
        add: 2,
        replace: 1,
        options: [
            {
                type: 'value',
                uuid: mockedUUID2,
                path: 'some.path',
                value: 'some value',
                name: 'My Label',
            },
        ],
        name: 'Some choice',
        description: 'Choice _desc_',
    };

    beforeAll(() => {
        fetch.mockImplementation(mockedApi({
            options: [{
                id: 1,
                name: "Example include",
                options: [],
                type: "multichoice",
            }],
        }));
        ListDataActions.fetchItems('options', 'data');
    });

    afterAll(() => fetch.resetMocks());

    beforeEach(() => {
        uuidv4.mockImplementation(
            jest.fn().mockReturnValueOnce(mockedUUID1)
        );
    });

    describe('when rendering', () => {
        const setState = jest.fn();

        it('should work with minimum props', () => {
            const tree = renderer.create(
                <MultichoiceOption
                    setState={setState}
                />
            ).toJSON();

            expect(tree).toMatchSnapshot();
        })

        it('should work with full props', () => {
            const tree = renderer.create(
                <MultichoiceOption
                    {...fullProps}
                    setState={setState}
                />
            ).toJSON();

            expect(tree).toMatchSnapshot();
        })
    });

    it('should emit changes in the label', () => {
        const setState = jest.fn();

        const wrapper = mount(
            <MultichoiceOption
                setState={setState}
            />
        );

        wrapper.find('input[placeholder="Label..."]').simulate('change', {
            target: { value: fullProps.name }
        });

        expect(setState).toBeCalledWith({
            type: 'multichoice',
            uuid: mockedUUID1,
            name: fullProps.name,
        });
    });

    it('should emit changes in the child component', () => {
        const setState = jest.fn();

        const wrapper = mount(
            <MultichoiceOption
                {...fullProps}
                setState={setState}
            />
        );

        wrapper.find('input[type="text"]').at(1).simulate('change', {
            target: { value: 'other.path' }
        });

        expect(setState).toBeCalledWith({
            type: 'multichoice',
            uuid: mockedUUID1,
            options: [
                {...fullProps.options[0], path: 'other.path' }
            ],
        });
    });
});
