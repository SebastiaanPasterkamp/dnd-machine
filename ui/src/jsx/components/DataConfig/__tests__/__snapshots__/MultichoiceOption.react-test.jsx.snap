// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MultichoiceOption when rendering should work with full props 1`] = `
<fieldset
  className="nice-form-group"
>
  <legend>
    Multichoice option
  </legend>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Label
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Label..."
      type="text"
      value="Some choice"
    />
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Description
    </span>
    <div
      className="markdown-textedit small preview"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <textarea
        className="nice-form-control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Description..."
        rows={8}
        value="Choice _desc_"
      />
      <span
        className="markdown-textedit__preview"
      >
        <p>
          Choice 
          <em>
            desc
          </em>
        </p>
      </span>
    </div>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Add
    </span>
    <input
      className="nice-form-control"
      disabled={false}
      min={0}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Add..."
      type="number"
      value={2}
    />
    <span
      className="nice-input-addon"
    >
      Limit
    </span>
    <input
      className="nice-form-control"
      disabled={true}
      min={0}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Limit..."
      type="number"
      value={0}
    />
    <span
      className="nice-input-addon"
    >
      Replace
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      min={0}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Replace..."
      type="number"
      value={1}
    />
  </div>
  <ul
    className="select-list-component"
  >
    <li>
      <fieldset
        className="nice-form-group"
      >
        <legend>
          Value option
        </legend>
        <div
          className="nice-control-group"
        >
          <span
            className="nice-input-addon"
          >
            Path
          </span>
          <input
            className="nice-form-control"
            disabled={undefined}
            onChange={[Function]}
            onKeyPress={null}
            placeholder="Path..."
            type="text"
            value="some.path"
          />
        </div>
        <input
          checked={false}
          className="nice-toggle"
          disabled={undefined}
          id="toggle-1"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="nice-toggle-btn"
          htmlFor="toggle-1"
        >
          Hidden
        </label>
        <div
          className="nice-control-group"
        >
          <span
            className="nice-input-addon"
          >
            Label
          </span>
          <input
            className="nice-form-control"
            disabled={undefined}
            onChange={[Function]}
            onKeyPress={null}
            placeholder="Label..."
            type="text"
            value="My Label"
          />
        </div>
        <div
          className="nice-control-group"
        >
          <span
            className="nice-input-addon"
          >
            Value
          </span>
          <div
            className="markdown-textedit small preview"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <textarea
              className="nice-form-control"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Value..."
              rows={8}
              value="some value"
            />
            <span
              className="markdown-textedit__preview"
            >
              <p>
                some value
              </p>
            </span>
          </div>
        </div>
      </fieldset>
      <button
        className="nice-btn select-list-component__delete warning icon fa-minus"
        name="del"
        onClick={[Function]}
      >
        â€‹
      </button>
    </li>
    <li
      className="select-list-component__last"
    >
      <div
        className="nice-dropdown nice-form-control select-list-component__add"
        color="primary"
        name="add"
      >
        <button
          className="nice-btn"
          disabled={false}
          onClick={[Function]}
        >
          Add...
          <i
            className="icon fa-angle-down"
          />
        </button>
        <ul
          className="dropdown-menu"
          onClick={[Function]}
        >
          <input
            className="nice-form-control"
            data-name="filter"
            disabled={undefined}
            onChange={[Function]}
            onClick={[Function]}
            onKeyPress={null}
            placeholder="Filter..."
            type="text"
            value=""
          />
          <li
            className="info"
            data-value={null}
            onClick={[Function]}
          >
            <a>
              Add...
            </a>
          </li>
          <li
            className={null}
            data-value="ability_score"
            onClick={[Function]}
          >
            <a>
              Ability Score Improvement
            </a>
          </li>
          <li
            className={null}
            data-value="dict"
            onClick={[Function]}
          >
            <a>
              Dictionary option
            </a>
          </li>
          <li
            className={null}
            data-value="choice"
            onClick={[Function]}
          >
            <a>
              Choice option
            </a>
          </li>
          <li
            className={null}
            data-value="config"
            onClick={[Function]}
          >
            <a>
              Config option
            </a>
          </li>
          <li
            className={null}
            data-value="list"
            onClick={[Function]}
          >
            <a>
              List option
            </a>
          </li>
          <li
            className={null}
            data-value="manual"
            onClick={[Function]}
          >
            <a>
              Manual value option
            </a>
          </li>
          <li
            className={null}
            data-value="multichoice"
            onClick={[Function]}
          >
            <a>
              Multichoice option
            </a>
          </li>
          <li
            className={null}
            data-value="objectlist"
            onClick={[Function]}
          >
            <a>
              Object list option
            </a>
          </li>
          <li
            className={null}
            data-value="value"
            onClick={[Function]}
          >
            <a>
              Value option
            </a>
          </li>
          <li>
            Prefab options
            <ul>
              <li
                className={null}
                data-value="asi"
                onClick={[Function]}
              >
                <a>
                  Ability Score Increase
                </a>
              </li>
              <li
                className={null}
                data-value="asi or feat"
                onClick={[Function]}
              >
                <a>
                  ASI or Feat
                </a>
              </li>
              <li
                className={null}
                data-value="darkvision"
                onClick={[Function]}
              >
                <a>
                  Darkvision
                </a>
              </li>
              <li
                className={null}
                data-value="equipment"
                onClick={[Function]}
              >
                <a>
                  Equipment
                </a>
              </li>
              <li
                className={null}
                data-value="equipment packs"
                onClick={[Function]}
              >
                <a>
                  Equipment Packs
                </a>
              </li>
              <li
                className={null}
                data-value="hit dice"
                onClick={[Function]}
              >
                <a>
                  Hit Dice
                </a>
              </li>
              <li
                className={null}
                data-value="personalities"
                onClick={[Function]}
              >
                <a>
                  Personalities
                </a>
              </li>
              <li>
                Proficiencies
                <ul>
                  <li
                    className={null}
                    data-value="proficiencies.armor"
                    onClick={[Function]}
                  >
                    <a>
                      Armor Proficiency
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.weapons"
                    onClick={[Function]}
                  >
                    <a>
                      Weapon Proficiency
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.languages"
                    onClick={[Function]}
                  >
                    <a>
                      Language Proficiency
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.saving_throws"
                    onClick={[Function]}
                  >
                    <a>
                      Saving Throws
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.expertise"
                    onClick={[Function]}
                  >
                    <a>
                      Expertise (Double Proficiency)
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.skills"
                    onClick={[Function]}
                  >
                    <a>
                      Skill Proficiency
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.talent"
                    onClick={[Function]}
                  >
                    <a>
                      Talent (Half Proficiency)
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.tools"
                    onClick={[Function]}
                  >
                    <a>
                      Tool Proficiency
                    </a>
                  </li>
                </ul>
              </li>
              <li
                className={null}
                data-value="size"
                onClick={[Function]}
              >
                <a>
                  Size
                </a>
              </li>
              <li
                className={null}
                data-value="speed"
                onClick={[Function]}
              >
                <a>
                  Speed
                </a>
              </li>
              <li>
                Spells
                <ul>
                  <li
                    className={null}
                    data-value="ability"
                    onClick={[Function]}
                  >
                    <a>
                      Spellcasting Ability
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="cantrips"
                    onClick={[Function]}
                  >
                    <a>
                      Cantrips
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="spell.list"
                    onClick={[Function]}
                  >
                    <a>
                      Spells
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="casting"
                    onClick={[Function]}
                  >
                    <a>
                      Casting Spells
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="focus"
                    onClick={[Function]}
                  >
                    <a>
                      Spellcasting Focus
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="preparing"
                    onClick={[Function]}
                  >
                    <a>
                      Preparing Spells
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="ritual"
                    onClick={[Function]}
                  >
                    <a>
                      Ritual Casting
                    </a>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </li>
  </ul>
</fieldset>
`;

exports[`Component: MultichoiceOption when rendering should work with minimum props 1`] = `
<fieldset
  className="nice-form-group"
>
  <legend>
    Multichoice option
  </legend>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Label
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Label..."
      type="text"
      value=""
    />
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Description
    </span>
    <div
      className="markdown-textedit small preview"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="markdown-textedit__placeholder"
      >
        Description...
      </span>
      <textarea
        className="nice-form-control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Description..."
        rows={8}
        value=""
      />
      <span
        className="markdown-textedit__preview"
      />
    </div>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Add
    </span>
    <input
      className="nice-form-control"
      disabled={false}
      min={0}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Add..."
      type="number"
      value={0}
    />
    <span
      className="nice-input-addon"
    >
      Limit
    </span>
    <input
      className="nice-form-control"
      disabled={false}
      min={0}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Limit..."
      type="number"
      value={0}
    />
    <span
      className="nice-input-addon"
    >
      Replace
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      min={0}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Replace..."
      type="number"
      value={0}
    />
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Include
    </span>
    <div
      className="nice-dropdown nice-form-control"
    >
      <button
        className="nice-btn"
        disabled={false}
        onClick={[Function]}
      >
        Include option...
        <i
          className="icon fa-angle-down"
        />
      </button>
      <ul
        className="dropdown-menu"
        onClick={[Function]}
      >
        <li
          className="info"
          data-value={null}
          onClick={[Function]}
        >
          <a>
            Include option...
          </a>
        </li>
        <li
          className={null}
          data-value={1}
          onClick={[Function]}
        >
          <a>
            Example include
          </a>
        </li>
      </ul>
    </div>
  </div>
  <ul
    className="select-list-component"
  >
    <li
      className="select-list-component__last"
    >
      <div
        className="nice-dropdown nice-form-control select-list-component__add"
        color="primary"
        name="add"
      >
        <button
          className="nice-btn"
          disabled={false}
          onClick={[Function]}
        >
          Add...
          <i
            className="icon fa-angle-down"
          />
        </button>
        <ul
          className="dropdown-menu"
          onClick={[Function]}
        >
          <input
            className="nice-form-control"
            data-name="filter"
            disabled={undefined}
            onChange={[Function]}
            onClick={[Function]}
            onKeyPress={null}
            placeholder="Filter..."
            type="text"
            value=""
          />
          <li
            className="info"
            data-value={null}
            onClick={[Function]}
          >
            <a>
              Add...
            </a>
          </li>
          <li
            className={null}
            data-value="ability_score"
            onClick={[Function]}
          >
            <a>
              Ability Score Improvement
            </a>
          </li>
          <li
            className={null}
            data-value="dict"
            onClick={[Function]}
          >
            <a>
              Dictionary option
            </a>
          </li>
          <li
            className={null}
            data-value="choice"
            onClick={[Function]}
          >
            <a>
              Choice option
            </a>
          </li>
          <li
            className={null}
            data-value="config"
            onClick={[Function]}
          >
            <a>
              Config option
            </a>
          </li>
          <li
            className={null}
            data-value="list"
            onClick={[Function]}
          >
            <a>
              List option
            </a>
          </li>
          <li
            className={null}
            data-value="manual"
            onClick={[Function]}
          >
            <a>
              Manual value option
            </a>
          </li>
          <li
            className={null}
            data-value="multichoice"
            onClick={[Function]}
          >
            <a>
              Multichoice option
            </a>
          </li>
          <li
            className={null}
            data-value="objectlist"
            onClick={[Function]}
          >
            <a>
              Object list option
            </a>
          </li>
          <li
            className={null}
            data-value="value"
            onClick={[Function]}
          >
            <a>
              Value option
            </a>
          </li>
          <li>
            Prefab options
            <ul>
              <li
                className={null}
                data-value="asi"
                onClick={[Function]}
              >
                <a>
                  Ability Score Increase
                </a>
              </li>
              <li
                className={null}
                data-value="asi or feat"
                onClick={[Function]}
              >
                <a>
                  ASI or Feat
                </a>
              </li>
              <li
                className={null}
                data-value="darkvision"
                onClick={[Function]}
              >
                <a>
                  Darkvision
                </a>
              </li>
              <li
                className={null}
                data-value="equipment"
                onClick={[Function]}
              >
                <a>
                  Equipment
                </a>
              </li>
              <li
                className={null}
                data-value="equipment packs"
                onClick={[Function]}
              >
                <a>
                  Equipment Packs
                </a>
              </li>
              <li
                className={null}
                data-value="hit dice"
                onClick={[Function]}
              >
                <a>
                  Hit Dice
                </a>
              </li>
              <li
                className={null}
                data-value="personalities"
                onClick={[Function]}
              >
                <a>
                  Personalities
                </a>
              </li>
              <li>
                Proficiencies
                <ul>
                  <li
                    className={null}
                    data-value="proficiencies.armor"
                    onClick={[Function]}
                  >
                    <a>
                      Armor Proficiency
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.weapons"
                    onClick={[Function]}
                  >
                    <a>
                      Weapon Proficiency
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.languages"
                    onClick={[Function]}
                  >
                    <a>
                      Language Proficiency
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.saving_throws"
                    onClick={[Function]}
                  >
                    <a>
                      Saving Throws
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.expertise"
                    onClick={[Function]}
                  >
                    <a>
                      Expertise (Double Proficiency)
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.skills"
                    onClick={[Function]}
                  >
                    <a>
                      Skill Proficiency
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.talent"
                    onClick={[Function]}
                  >
                    <a>
                      Talent (Half Proficiency)
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="proficiencies.tools"
                    onClick={[Function]}
                  >
                    <a>
                      Tool Proficiency
                    </a>
                  </li>
                </ul>
              </li>
              <li
                className={null}
                data-value="size"
                onClick={[Function]}
              >
                <a>
                  Size
                </a>
              </li>
              <li
                className={null}
                data-value="speed"
                onClick={[Function]}
              >
                <a>
                  Speed
                </a>
              </li>
              <li>
                Spells
                <ul>
                  <li
                    className={null}
                    data-value="ability"
                    onClick={[Function]}
                  >
                    <a>
                      Spellcasting Ability
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="cantrips"
                    onClick={[Function]}
                  >
                    <a>
                      Cantrips
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="spell.list"
                    onClick={[Function]}
                  >
                    <a>
                      Spells
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="casting"
                    onClick={[Function]}
                  >
                    <a>
                      Casting Spells
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="focus"
                    onClick={[Function]}
                  >
                    <a>
                      Spellcasting Focus
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="preparing"
                    onClick={[Function]}
                  >
                    <a>
                      Preparing Spells
                    </a>
                  </li>
                  <li
                    className={null}
                    data-value="ritual"
                    onClick={[Function]}
                  >
                    <a>
                      Ritual Casting
                    </a>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </li>
  </ul>
</fieldset>
`;
